File: backend/registrations/registerParticipant.mjs
import { PutCommand, QueryCommand, GetCommand } from "@aws-sdk/lib-dynamodb";
import { SQSClient, SendMessageCommand } from "@aws-sdk/client-sqs";
import { db, EVENTS_TABLE, REGISTRATIONS_TABLE } from "./shared/db.mjs";
import { ok, badRequest, notFound, conflict, serverError, cors } from "./shared/response.mjs";
import { newId, registrationNumber } from "./shared/ids.mjs";
import { audit } from "./shared/auth.mjs";

const sqs = new SQSClient({});

function validateEmail(email) {
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
}

export async function handler(event) {
  if (event.httpMethod === "OPTIONS") return cors();

  const { eventId } = event.pathParameters ?? {};
  if (!eventId) return badRequest("eventId is required");

  let body;
  try {
    body = JSON.parse(event.body ?? "{}");
  } catch {
    return badRequest("Invalid JSON");
  }

  const { fullName, organisation, email, phone, position } = body;
  if (!fullName || fullName.trim().length < 2) return badRequest("fullName is required");
  if (!organisation || organisation.trim().length < 2) return badRequest("organisation is required");
  if (!email || !validateEmail(email.trim())) return badRequest("Valid email is required");

  // Check event exists and registration is open
  const eventResult = await db.send(new GetCommand({ TableName: EVENTS_TABLE, Key: { eventId } }));
  if (!eventResult.Item) return notFound("Event not found");
  if (!eventResult.Item.registrationOpen) return badRequest("Registration is currently closed");

  // Check for duplicate email in this event
  const existing = await db.send(new QueryCommand({
    TableName: REGISTRATIONS_TABLE,
    IndexName: "email-eventId-index",
    KeyConditionExpression: "email = :email AND eventId = :eventId",
    ExpressionAttributeValues: { ":email": email.toLowerCase().trim(), ":eventId": eventId },
    Limit: 1,
  }));
  if ((existing.Items ?? []).length > 0) return conflict("This email is already registered for this event");

  const regNumber = registrationNumber(
    eventResult.Item.name.replace(/\s+/g, "").slice(0, 6).toUpperCase()
  );

  const item = {
    registrationId: newId(),
    eventId,
    registrationNumber: regNumber,
    fullName: fullName.trim(),
    organisation: organisation.trim(),
    email: email.toLowerCase().trim(),
    phone: phone?.trim() ?? null,
    position: position?.trim() ?? null,
    registrationType: "online",
    checkedInAt: null,
    checkedInBy: null,
    badgePrintedAt: null,
    badgePrintCount: 0,
    createdBy: null,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
  };

  try {
    await db.send(new PutCommand({ TableName: REGISTRATIONS_TABLE, Item: item }));
    await audit("participant.registered", {
      entity: "participant",
      entityId: item.registrationId,
      meta: { name: item.fullName, email: item.email, phone: item.phone, reg: item.registrationNumber, eventId },
    });

    // Enqueue post-registration processing (email, audit) via SQS
    if (process.env.REGISTRATION_QUEUE_URL) {
      await sqs.send(new SendMessageCommand({
        QueueUrl: process.env.REGISTRATION_QUEUE_URL,
        MessageBody: JSON.stringify({
          registrationId: item.registrationId,
          email: item.email,
          fullName: item.fullName,
          registrationNumber: item.registrationNumber,
          eventName: eventResult.Item.name,
          eventDate: eventResult.Item.date,
          venue: eventResult.Item.venue,
        }),
      }));
    }

    return ok({ registrationNumber: item.registrationNumber, registrationId: item.registrationId });
  } catch (e) {
    console.error(e);
    return serverError();
  }
}

File: backend/registrations/registerParticipant.py
import json
import logging
import os
import re

import boto3

from shared.auth import audit, iso_now
from shared.db import events_table, registrations_table
from shared.ids import new_id, registration_number
from shared.response import bad_request, conflict, cors, not_found, ok, server_error, with_cors

log = logging.getLogger()

EMAIL_RE = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")

_sns = None


def _sns_client():
    global _sns
    if _sns is None:
        _sns = boto3.client("sns")
    return _sns


def _valid_email(email):
    return bool(EMAIL_RE.match(email))


@with_cors
def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return cors()

    event_id = (event.get("pathParameters") or {}).get("eventId")
    if not event_id:
        return bad_request("eventId is required")

    try:
        body = json.loads(event.get("body") or "{}")
    except (ValueError, TypeError):
        return bad_request("Invalid JSON")

    full_name = body.get("fullName")
    organisation = body.get("organisation")
    email = body.get("email")
    phone = body.get("phone")
    position = body.get("position")

    if not full_name or len(full_name.strip()) < 2:
        return bad_request("fullName is required")
    if not organisation or len(organisation.strip()) < 2:
        return bad_request("organisation is required")
    if not email or not _valid_email(email.strip()):
        return bad_request("Valid email is required")

    email_norm = email.strip().lower()

    # Check event exists and registration is open
    event_item = events_table().get_item(Key={"eventId": event_id}).get("Item")
    if not event_item:
        return not_found("Event not found")
    if not event_item.get("registrationOpen"):
        return bad_request("Registration is currently closed")

    # Check for duplicate email in this event
    existing = registrations_table().query(
        IndexName="email-eventId-index",
        KeyConditionExpression="email = :email AND eventId = :eventId",
        ExpressionAttributeValues={":email": email_norm, ":eventId": event_id},
        Limit=1,
    )
    if existing.get("Items"):
        return conflict("This email is already registered for this event")

    prefix = re.sub(r"\s+", "", event_item.get("name") or "")[:6].upper()
    reg_number = registration_number(prefix)
    now = iso_now()

    item = {
        "registrationId": new_id(),
        "eventId": event_id,
        "registrationNumber": reg_number,
        "fullName": full_name.strip(),
        "organisation": organisation.strip(),
        "email": email_norm,
        "phone": phone.strip() if phone else None,
        "position": position.strip() if position else None,
        "registrationType": "online",
        "checkedInAt": None,
        "checkedInBy": None,
        "badgePrintedAt": None,
        "badgePrintCount": 0,
        "createdBy": None,
        "createdAt": now,
        "updatedAt": now,
    }

    try:
        registrations_table().put_item(Item=item)
        audit("participant.registered", entity="participant", entity_id=item["registrationId"],
              meta={"name": item["fullName"], "email": item["email"], "phone": item["phone"],
                    "reg": item["registrationNumber"], "eventId": event_id})

        # Publish post-registration processing (confirmation email) to SNS;
        # sendConfirmationEmail consumes the topic. Best-effort: a publish
        # failure must not fail the registration itself.
        topic_arn = os.environ.get("SNS_TOPIC_ARN")
        if topic_arn:
            try:
                _sns_client().publish(
                    TopicArn=topic_arn,
                    Message=json.dumps({
                        "registrationId": item["registrationId"],
                        "email": item["email"],
                        "fullName": item["fullName"],
                        "registrationNumber": item["registrationNumber"],
                        "eventName": event_item.get("name"),
                        "eventDate": event_item.get("date"),
                        "venue": event_item.get("venue"),
                    }),
                )
            except Exception:
                log.exception("SNS publish failed for %s", item["registrationId"])

        return ok({"registrationNumber": item["registrationNumber"],
                   "registrationId": item["registrationId"]})
    except Exception:
        log.exception("registerParticipant failed")
        return server_error()
